src/offset_log.rs

```diff
@@ -234,100 +151,168 @@ pub struct Data {
     pub id: u64,
 }
 
+fn size_of_frame_tail<T>() -> usize {
+    size_of::<u32>() + size_of::<T>()
+}
 fn size_of_framing_bytes<T>() -> usize {
     size_of::<u32>() * 2 + size_of::<T>()
 }
 
-fn is_valid_frame<T>(buf: &BytesMut, data_size: usize) -> bool {
-    let second_data_size_index = data_size + size_of::<u32>();
+pub fn encode<T>(offset: u64, item: &[u8], dest: &mut BytesMut) -> Result<u64, Error> {
+    let chunk_size = size_of_framing_bytes::<T>() + item.len();
+    dest.reserve(chunk_size);
+    dest.put_u32_be(item.len() as u32);
+    dest.put_slice(&item);
+    dest.put_u32_be(item.len() as u32);
+    let new_offset = offset + chunk_size as u64;
+    // self.length += chunk_size as u64;
+
+    dest.put_uint_be(new_offset, size_of::<T>());
+    Ok(new_offset)
+}
+
+pub fn validate_entry<T>(offset: usize, data_size: usize, rest: &[u8]) -> Result<u64, Error> {
```

Comment:
Just thinking about `offset: usize`
I think we should make this u64 otherwise 32bit systems can't handle offset files larger than 4GB.
It's probably my mistake that you've inherited. Doesn't need to be fixed now.

Reply:
Whoops, just a mistake. All of the other offsets are u64, so that one should be too.

Reply:
Speaking of which, the usize vs u64 business is a bit of a mess. There are a bunch of `as u64` and `as usize`s sprinkled about. I'll clean it up.